src/webgpu/shader/execution/expression/expression.ts

```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
       // Constant eval
       //////////////////////////////////////////////////////////////////////////
       let body = '';
-      if (parameterTypes.some(ty => scalarTypeOf(ty).kind === 'abstract-float')) {
+      if (
+        scalarTypeOf(resultType).kind !== 'abstract-float' &&
+        parameterTypes.some(ty => scalarTypeOf(ty).kind === 'abstract-float')
+      ) {
         // Directly assign the expression to the output, to avoid an
         // intermediate store, which will concretize the value early
         body = cases
           .map(
             (c, i) => `  outputs[${i}].value = ${expressionBuilder(map(c.input, v => v.wgsl()))};`
           )
           .join('\n  ');
+      } else if (scalarTypeOf(resultType).kind === 'abstract-float') {
+        // AbstractFloats are f64s under the hood. WebGPU does not support
+        // putting f64s in buffers, so the result needs to be split up into u32s
+        // and rebuilt in the test framework.
+        // This is complicated by the fact that user defined functions cannot
+        // take/return AbstractFloats, and AbstractFloats cannot be stored in
+        // variables, so the code cannot just inject a simple utility function
+        // at the top of the shader, instead this snippet needs to be inlined
+        // everywhere the test needs to return an AbstractFloat.
+        //
+        // Documentation of the snippet out of line, since shader length affects
+        // compilation time significantly on some backends.
+        //
+        //   const exponent_bias = 1022;
+        //   // Use frexp() to obtain the exponent and fractional parts
+        //   const f = frexp(abs(${expressionBuilder(map(c.input, v => v.wgsl()))}));
+        //
+        //   // Zero behaves like a subnormal, so needs special casing later
+        //   const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
+        //   const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
+        //
+        //   // MSB of the high_u32 is 1 if the value is negative, otherwise 0
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expressionBuilder(map(c.input, v => v.wgsl()))} < 0);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special case for zero
+        //   const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, special case for zero
+        //   const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
+        //
+        //   // Subtract out the high 20 bits from the mantissa.
+        //   // const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
```

Comment:
Why is this using 1 << 20 when the upper bits are extracted with 1 << 21 ?


This code could use an example value as an illustration.  e.g. 1.9abcdefp+1

Reply:
Also, I recommend adding unit tests.